iwyu.cc

```diff
@@ -2560,7 +2553,24 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   }
 
   void ReportTypeUseInternal(SourceLocation used_loc, const Type* type,
-                             const set<const Type*>& blocked_types) {
+                             const set<const Type*>& blocked_types,
+                             UseKind use_kind) {
+    type = Desugar(type);
+    if (use_kind == UseKind::RemoveRefs ||
+        use_kind == UseKind::RemoveRefsAndPtr) {
+      if (const auto* ref_type = dyn_cast_or_null<ReferenceType>(type))
```

Comment:
I'm guessing this cast is necessary (rather than `type->isReferenceType()`) because you want to `getPointeeTypeAsWritten()`?

....

Reply:
Both `isReferenceType()` and `isPointerType()` check canonical type directly, through all the `TypedefType`s that may occur in the sugar chain. It should be avoided here because `TypedefType`s should be "caught" and handled here (in contrast to `InstantiatedTemplateVisitor` implementation of `ReportTypeUse`).

Reply:
Aah, subtle difference. Could you add an educational comment about that here?